account_payment_partner/wizard/payment_order_create.py

```diff
@@ -30,8 +30,25 @@ class PaymentOrderCreate(models.TransientModel):
     def extend_payment_order_domain(self, payment_order, domain):
         res = super(PaymentOrderCreate, self).extend_payment_order_domain(
             payment_order, domain)
-        domain += ['|', '|',
-                   ('invoice', '=', False),
+        # Monkey patch for fixing problem with the core search function
+        # when args has ('invoice', '=', False), referred in the issue #4857
+        # (https://github.com/odoo/odoo/issues/4857)
+        #
+        # Original domain:
+        # domain += ['|', '|',
+        #            ('invoice', '=', False),
+        #            ('invoice.payment_mode_id', '=', False),
+        #            ('invoice.payment_mode_id', '=', payment_order.mode.id)]
+        self.env.cr.execute(
+            'SELECT l.id '
+            'FROM account_move_line l '
+            'LEFT OUTER JOIN account_invoice i '
+            'ON l.move_id = i.move_id '
+            'WHERE i.id IS NULL')
```

Comment:
Well, this patch avoid the problem simply storing the field in DB and making regular searches. I can completely overwrite the invoice field, but that was one of the things we try to avoid in the past. Doesn't it?